### DISCUSSION TITLE
如何修改umi.ts入口文件的代码

### DISCUSSION THREAD

**Discussion by mutour123** [umijs/umi#12278] [Q&A]
使用无界微前端时，需要将umi实现的应用作为子应用嵌入到其他应用里，无界推荐使用单例模式，需要手动写生命周期，如下：
if (window.__POWERED_BY_WUJIE__) {
  window.__WUJIE_MOUNT = () => {
    ReactDOM.render(
      <React.StrictMode>
        <App />
      </React.StrictMode>,
      document.getElementById("root")
    );
  };
  window.__WUJIE_UNMOUNT = () => {
    ReactDOM.unmountComponentAtNode(document.getElementById("root"));
  };
} else {
  ReactDOM.render(
    <React.StrictMode>
      <App />
    </React.StrictMode>,
    document.getElementById("root")
  );
}

umi生成的入口文件为：

![image](https://github.com/umijs/umi/assets/19345200/52c39ac2-9fde-4c04-9c8c-3da786e448af)

请问怎么可以修改这个入口文件，以满足上诉需求

**Comment by fz6m** [comment]
如果你要修改 webpack 配置，就用 [`chainWebpack`](https://umijs.org/docs/api/plugin-api#chainwebpack) 这个配置。

如果要操作 react 挂载的 root 节点，就导入 `__getRoot` 这个方法获取（正常情况 umi 不暴露渲染的根节点，这个方法只用在特殊需要的场景）。

如果要操作应用入口（在 umi 里应用的入口是 `src/.umi/umi.ts` ，其他应用可能是 `src/index.tsx` 等）导出的内容，可以参考 umi [qiankun slave](https://github.com/umijs/umi/blob/master/packages/plugins/src/qiankun/slave.ts) 子应用的插件实现，其中通过调用插件 api 插入代码块到 `umi.ts` 来做到。

如果要控制应用 render 的时机，可以在应用的运行时配置文件 `src/app.ts` 里导出 render 来控制：

```ts
// src/app.ts

export const render = (oldRender: any) => {
  // ↓ 在准备好后，需要的时机再开始渲染
  oldRender()
}
```

如果要适配其他的微前端框架，建议还是写一个插件吧，参考 umi [qiankun 插件](https://github.com/umijs/umi/blob/master/packages/plugins/src/qiankun.ts) 即可。
